--- a/src/cursor.cpp
+++ b/src/cursor.cpp
@@ -24,14 +24,7 @@
 #include "getdata.h"
 #include "dbspecific.h"
 #include <datetime.h>
-
-enum
-{
-    CURSOR_REQUIRE_CNXN    = 0x00000001,
-    CURSOR_REQUIRE_OPEN    = 0x00000003, // includes _CNXN
-    CURSOR_REQUIRE_RESULTS = 0x00000007, // includes _OPEN
-    CURSOR_RAISE_ERROR     = 0x00000010,
-};
+#include "npcontainer.h"

 inline bool StatementIsValid(Cursor* cursor)
 {
@@ -45,7 +38,7 @@ inline bool Cursor_Check(PyObject* o)
     return o != 0 && Py_TYPE(o) == &CursorType;
 }

-static Cursor* Cursor_Validate(PyObject* obj, DWORD flags)
+Cursor* Cursor_Validate(PyObject* obj, DWORD flags)
 {
     //  Validates that a PyObject is a Cursor (like Cursor_Check) and optionally some other requirements controlled by
     //  `flags`.  If valid and all requirements (from the flags) are met, the cursor is returned, cast to Cursor*.
@@ -2396,6 +2389,7 @@ static PyMethodDef Cursor_methods[] =
     { "fetchone",         (PyCFunction)Cursor_fetchone,         METH_NOARGS,                fetchone_doc         },
     { "fetchall",         (PyCFunction)Cursor_fetchall,         METH_NOARGS,                fetchall_doc         },
     { "fetchmany",        (PyCFunction)Cursor_fetchmany,        METH_VARARGS,               fetchmany_doc        },
+    { "fetchdictarray",   (PyCFunction)Cursor_fetchdictarray,   METH_VARARGS|METH_KEYWORDS, fetchdictarray_doc   },
     { "nextset",          (PyCFunction)Cursor_nextset,          METH_NOARGS,                nextset_doc          },
     { "tables",           (PyCFunction)Cursor_tables,           METH_VARARGS|METH_KEYWORDS, tables_doc           },
     { "columns",          (PyCFunction)Cursor_columns,          METH_VARARGS|METH_KEYWORDS, columns_doc          },
diff --git a/src/cursor.h b/src/cursor.h
index 657eb48..d7a5050 100644
--- a/src/cursor.h
+++ b/src/cursor.h
@@ -13,6 +13,14 @@
 #ifndef CURSOR_H
 #define CURSOR_H

+enum
+{
+    CURSOR_REQUIRE_CNXN    = 0x00000001,
+    CURSOR_REQUIRE_OPEN    = 0x00000003, // includes _CNXN
+    CURSOR_REQUIRE_RESULTS = 0x00000007, // includes _OPEN
+    CURSOR_RAISE_ERROR     = 0x00000010,
+};
+
 struct Connection;

 struct ColumnInfo
@@ -162,5 +170,6 @@ void Cursor_init();

 Cursor* Cursor_New(Connection* cnxn);
 PyObject* Cursor_execute(PyObject* self, PyObject* args);
+Cursor* Cursor_Validate(PyObject* obj, DWORD flags);

 #endif
--
2.45.2
