--- a/src/pyodbcmodule.cpp
+++ b/src/pyodbcmodule.cpp
@@ -95,7 +95,7 @@ struct ExcInfo
     const char* szDoc;
 };

-#define MAKEEXCINFO(name, parent, doc) { #name, "pyodbc." #name, &name, &parent, doc }
+#define MAKEEXCINFO(name, parent, doc) { #name, "npyodbc." #name, &name, &parent, doc }

 static ExcInfo aExcInfos[] = {
     MAKEEXCINFO(Error, PyExc_Exception,
@@ -174,7 +174,7 @@ PyObject* GetClassForThread(const char* szModule, const char* szClass)
     {
         // I don't know why there wouldn't be thread state so I'm going to raise an exception
         // unless I find more info.
-        return PyErr_Format(PyExc_Exception, "pyodbc: PyThreadState_GetDict returned NULL");
+        return PyErr_Format(PyExc_Exception, "npyodbc: PyThreadState_GetDict returned NULL");
     }

     // Check the cache.  GetItemString returns a borrowed reference.
@@ -298,7 +298,7 @@ static bool AllocateEnv()

     if (!SQL_SUCCEEDED(SQLAllocHandle(SQL_HANDLE_ENV, SQL_NULL_HANDLE, &henv)))
     {
-        PyErr_SetString(PyExc_RuntimeError, "Can't initialize module pyodbc.  SQLAllocEnv failed.");
+        PyErr_SetString(PyExc_RuntimeError, "Can't initialize module npyodbc.  SQLAllocEnv failed.");
         return false;
     }

@@ -703,7 +703,7 @@ static char connect_doc[] =
     "The connection string will be passed to SQLDriverConnect, so a DSN connection\n"
     "can be created using:\n"
     "\n"
-    "  cnxn = pyodbc.connect('DSN=DataSourceName;UID=user;PWD=password')\n"
+    "  cnxn = npyodbc.connect('DSN=DataSourceName;UID=user;PWD=password')\n"
     "\n"
     "To connect without requiring a DSN, specify the driver and connection\n"
     "information:\n"
@@ -727,7 +727,7 @@ static char connect_doc[] =
     "\n"
     "Special Keywords\n"
     "\n"
-    "The following specal keywords are processed by pyodbc and are not added to the\n"
+    "The following specal keywords are processed by npyodbc and are not added to the\n"
     "connection string.  (If you must use these in your connection string, pass them\n"
     "as a string, not as keywords.)\n"
     "\n"
@@ -1148,7 +1148,7 @@ static bool CreateExceptions()

 static struct PyModuleDef moduledef = {
     PyModuleDef_HEAD_INIT,
-    "pyodbc",                   // m_name
+    "npyodbc",                   // m_name
     module_doc,
     -1,                         // m_size
     pyodbc_methods,             // m_methods
--
2.45.2
