diff --git a/src/App.css b/src/App.css
index b9d355d..dd25abf 100644
--- a/src/App.css
+++ b/src/App.css
@@ -18,6 +18,11 @@
   filter: drop-shadow(0 0 2em #61dafbaa);
 }
 
+.greet {
+  font-size: 1.5em;
+  font-weight: 500;
+}
+
 @keyframes logo-spin {
   from {
     transform: rotate(0deg);
diff --git a/src/App.jsx b/src/App.jsx
index f67355a..f139c05 100644
--- a/src/App.jsx
+++ b/src/App.jsx
@@ -4,7 +4,12 @@ import viteLogo from '/vite.svg'
 import './App.css'
 
 function App() {
-  const [count, setCount] = useState(0)
+  const [greet, setgreet] = useState("")
+
+  async function callGreet() {
+      let res = await window.paneer.invoke("greet", {});
+      setgreet(res);
+  }
 
   return (
     <>
@@ -16,11 +21,14 @@ function App() {
           <img src={reactLogo} className="logo react" alt="React logo" />
         </a>
       </div>
-      <h1>Vite + React</h1>
+      <h1>Vite + React + Paneer</h1>
       <div className="card">
-        <button onClick={() => setCount((count) => count + 1)}>
-          count is {count}
+        <button onClick={callGreet}>
+          Greet!
         </button>
+        <p className='greet'>
+          {greet}
+        </p>
         <p>
           Edit <code>src/App.jsx</code> and save to test HMR
         </p>
diff --git a/vite.config.js b/vite.config.js
index 8b0f57b..0a5127e 100644
--- a/vite.config.js
+++ b/vite.config.js
@@ -4,4 +4,8 @@ import react from '@vitejs/plugin-react'
 // https://vite.dev/config/
 export default defineConfig({
   plugins: [react()],
-  base: './'
+  base: './',
+  server: {
+    host: '127.0.0.1',
+    port: 5173,
+  }
 })
